In [1]:
# %tensorflow_version 1.x
# !pip install -q gpt-2-simple
import gpt_2_simple as gpt2
from datetime import datetime
# from google.colab import files
import tensorflow as tf
import tensorflow_hub as hub

# import shutil

from sklearn.model_selection import train_test_split
import pandas as pd
import numpy as np

import bert
from bert import run_classifier
from bert import optimization
from bert import tokenization

tf.logging.set_verbosity(tf.logging.ERROR)

/dhlabdata4/bborges/.conda/envs/bertp2/lib/python3.6/site-packages/tensorflow/python/framework/dtypes.py:516: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint8 = np.dtype([("qint8", np.int8, 1)])
/dhlabdata4/bborges/.conda/envs/bertp2/lib/python3.6/site-packages/tensorflow/python/framework/dtypes.py:517: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_quint8 = np.dtype([("quint8", np.uint8, 1)])
/dhlabdata4/bborges/.conda/envs/bertp2/lib/python3.6/site-packages/tensorflow/python/framework/dtypes.py:518: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint16 = np.dtype([("qint16", np.int16, 1)])
/dhlabdata4/bborges/.conda/envs/be

In [2]:
import csv
import sys
csv.field_size_limit(100000000)

131072

In [3]:
#@markdown Whether or not to finetune GPT-2
FINETUNE_GPT2 = False #@param {type:"boolean"}

In [3]:
from tensorflow.python.client import device_lib 
print(device_lib.list_local_devices())

[name: "/device:CPU:0"
device_type: "CPU"
memory_limit: 268435456
locality {
}
incarnation: 6181889679292703173
, name: "/device:XLA_CPU:0"
device_type: "XLA_CPU"
memory_limit: 17179869184
locality {
}
incarnation: 16176760870079880775
physical_device_desc: "device: XLA_CPU device"
]


In [2]:
# from google.colab import drive
# drive.mount('/content/drive', force_remount=True)

Mounted at /content/drive


In [5]:
if FINETUNE_GPT2:
    gpt2.download_gpt2(model_name="355M")

Fetching checkpoint: 1.05Mit [00:00, 505Mit/s]                                                      
Fetching encoder.json: 1.05Mit [00:00, 104Mit/s]                                                    
Fetching hparams.json: 1.05Mit [00:00, 737Mit/s]                                                    
Fetching model.ckpt.data-00000-of-00001: 1.42Git [00:05, 247Mit/s]                                  
Fetching model.ckpt.index: 1.05Mit [00:00, 269Mit/s]                                                
Fetching model.ckpt.meta: 1.05Mit [00:00, 85.4Mit/s]                                                
Fetching vocab.bpe: 1.05Mit [00:00, 109Mit/s]                                                       


In [6]:
if FINETUNE_GPT2:
#     file_name='bios_old_dates.csv'
#     shutil.copyfile("/content/drive/Shared drives/FDH/data/" + file_name, file_name)

    # don't be proprietary folks
    # gpt2.copy_file_from_gdrive(file_name)

    tf.reset_default_graph()
    sess = gpt2.start_tf_sess()
    gpt2.finetune(sess,
                dataset=file_name,
                model_name='355M',
                steps=2000,
                restore_from='latest',
                run_name='old_dates',
                print_every=10,
                sample_every=200,
                save_every=500
                )


Instructions for updating:
Use tf.where in 2.0, which has the same broadcast rule as np.where
Instructions for updating:
Please use tensorflow.python.ops.op_selector.get_backward_walk_ops.
Loading checkpoint models/355M/model.ckpt
INFO:tensorflow:Restoring parameters from models/355M/model.ckpt


100%|██████████| 1/1 [00:00<00:00, 22.10it/s]

Loading dataset...


dataset has 566149 tokens
Training...
[10 | 23.88] loss=3.27 avg=3.27
[20 | 38.97] loss=3.14 avg=3.21
[30 | 54.29] loss=2.86 avg=3.09
[40 | 69.82] loss=2.88 avg=3.04
[50 | 85.62] loss=2.89 avg=3.01
[60 | 101.59] loss=3.04 avg=3.01
[70 | 117.75] loss=3.17 avg=3.04
[80 | 133.91] loss=2.38 avg=2.95
[90 | 149.92] loss=3.30 avg=2.99
[100 | 165.89] loss=2.78 avg=2.97
[110 | 181.93] loss=2.66 avg=2.94
[120 | 198.01] loss=2.91 avg=2.94
[130 | 214.08] loss=3.22 avg=2.96
[140 | 230.13] loss=3.46 avg=3.00
[150 | 246.19] loss=2.75 avg=2.98
[160 | 262.24] loss=2.62 avg=2.96
[170 | 278.27] loss=2.70 avg=2.94
[180 | 294.32] loss=2.98 avg=2.94
[190 | 310.35] loss=2.96 avg=2.94
[200 | 326.37] loss=2.65 avg=2.93
======== SAMPLE 1 ========
 of God. After the death of Philip II in 1516, he and his brother Philip I, Prince of Austria, united the royal domains of Croatia, Montenegro and Serbia into one, following a long and painful struggle. Their victory over the Ottoman sultan, Mehmed II Mustafa, was a wa

In [9]:
# if FINETUNE_GPT2:
#     gpt2.copy_checkpoint_to_gdrive(run_name='old_dates')

In [6]:
# don't be proprietary folks
# gpt2.copy_file_from_gdrive("checkpoint_mediumgpt2_venetianscsv.tar")

# if not FINETUNE_GPT2:
#     import shutil
#     file_name='checkpoint_old_dates.tar'#mediumgpt2_sep_venetianscsv.tar'
#     shutil.copyfile("/content/drive/Shared drives/FDH/model/" + file_name, file_name)

In [7]:
# !tar xf checkpoint_old_dates.tar

In [30]:
if not FINETUNE_GPT2:
    sess = gpt2.start_tf_sess()
    gpt2.load_gpt2(sess, run_name='old_dates')

Loading checkpoint checkpoint/old_dates/model-2000


In [31]:
def generate_fake_bio(prompt):
    proposed_bio = str(gpt2.generate(sess, run_name='old_dates', return_as_list=True, 
                                  prefix=prompt, length=4096)[0])

    split_proposed_bio = proposed_bio.split('<|')[0]
    if split_proposed_bio.count('[SEP]') > 1:
        split_proposed_bio = ''.join(''.join(split_proposed_bio.split('[SEP]')[:2]).split('.')[:-2])
        split_proposed_bio = '.'.join(split_proposed_bio.split('.')[:-1]) + '.'

    return split_proposed_bio

In [6]:
print(generate_fake_bio('Giovanni Domenico was a Venetian opera singer. [SEP]'))

Giovanni Domenico was a Venetian opera singer. [SEP] He was the son and partner of composer Andrea Mantegna, and composed operas in the style of his father. He was the first in the family to graduate from the minor to the major opera. The Mantegna family was important in the creation of Venetian opera. Giovanni was the first to use the Mantuan school, and is considered to be the father of the Venetian opera.     Giovanni was the first to use the Mantuan school, and is considered to be the father of the Venetian opera. Giovanni Mantegna was born in Venice on 12 October 1609 into the noble family of Mantua. He began his musical education in the Dominican Order, and was educated at the convent of Santa Maria della Salute in Padua, his birthplace. He went to Venice, where he sang at the Scuola di San Giovanni Evangelista, and later moved to the Accademia, where he acquired a sister. He moved to Paris, where he sang at the Chambre, and then to Bologna, where he sang at the Gregoria.    He m

In [32]:
first_names = ['Achille', 'Adamo', 'Adelmo', 'Adriano', 'Agnolo', 'Agostino', 'Alberico', 'Alberto', 'Alderano', 'Aldo', 'Alessandro', 'Alessio', 'Alfio', 'Alfredo', 'Alphons', 'Amadeo', 'Amedeo', 'Amico', 'Amleto', 'Angelo', 'Annibale', 'Ansaldo', 'Antonello', 'Antonino', 'Antonio', 'Armando', 'Arnaldo', 'Arnulfo', 'Arsenio', 'Arturo', 'Atenolfo', 'Atenulfo', 'Augusto', 'Azeglio', 'Baccio', 'Baldassare', 'Bartolomeo', 'Benedetto', 'Benito', 'Benvenuto', 'Beppe', 'Bernardo', 'Biagio', 'Bruno', 'Calcedonio', 'Calogero', 'Camillo', 'Carlo', 'Carlo', 'Carmelo', 'Carmine', 'Cesare', 'Cipriano', 'Cirillo', 'Ciro', 'Claudio', 'Coluccio', 'Constanzo', 'Coriolano', 'Corrado', 'Costantino', 'Costanzo', 'Danilo', 'Damiano', 'Daniele', 'Daniello', 'Dante', 'Dario', 'Davide', 'Delfino', 'Dino', 'Dionigi', 'Domenico', 'Donatello', 'Donato', 'Durante', 'Edgardo', 'Edoardo', 'Elladio', 'Elmo', 'Emilio', 'Ennio', 'Enrico', 'Enzio', 'Enzo', 'Eraldo', 'Ermanno', 'Ermenegildo', 'Ermes', 'Ernesto', 'Ettore', 'Ezio', 'Fabio', 'Fabrizio', 'Fausto', 'Fedele', 'Federico', 'Federigo', 'Ferdinando', 'Filippo', 'Fiorenzo', 'Fiorino', 'Flavio', 'Francesco', 'Franco', 'Fredo', 'Fulvio', 'Gabriele', 'Gaetano', 'Galasso', 'Gaspare', 'Gastone', 'Gavino', 'Gennaro', 'Geppetto', 'Geronimo', 'Giacinto', 'Giacobbe', 'Giacomo', 'Giampaolo', 'Giampiero', 'Gian', 'Gian Carlo', 'Gianantonio', 'Giancarlo', 'Gianfrancesco', 'Gianfranco', 'Gianluca', 'Gianluigi', 'Gianmarco', 'Giannantonio', 'Gianni', 'Gianpaolo', 'Gianpietro', 'Gilberto', 'Gino', 'Gioacchino', 'Gioachino', 'Gioele', 'Gioffre', 'Gionata', 'Giordano', 'Giorgio', 'Giosuè', 'Giovanni', 'Giovanni Battista', 'Girolamo', 'Giuliano', 'Giulio', 'Giuseppe', 'Giustino', 'Goffredo', 'Graziano', 'Greco', 'Guarino', 'Guglielmo', 'Guido', 'Gustavo', 'Hugo', 'Ignazio', 'Ippazio', 'Ivan', 'Ivo', 'Jacopo', 'Lamberto', 'Lando', 'Laureano', 'Lazzaro', 'Leonardo', 'Leone', 'Leopoldo', 'Liberato', 'Liberto', 'Livio', 'Lodovico', 'Lorenzo', 'Lotario', 'Luca', 'Luchino', 'Luciano', 'Lucio', 'Ludovico', 'Luigi', 'Manuel', 'Marcantonio', 'Marcello', 'Marco', 'Mariano', 'Mario', 'Martino', 'Martino', 'Massimiliano', 'Massimo', 'Matteo', 'Mattia', 'Maurilio', 'Maurizio', 'Mauro', 'Michelangelo', 'Michele', 'Micheletto', 'Michelotto', 'Milo', 'Mirco', 'Mirko', 'Moreno', 'Nanni', 'Napoleone', 'Niccolò', 'Nico', 'Nicola', 'Nicolò', 'Nino', 'Nunzio', 'Omero', 'Orazio', 'Oreste', 'Orlando', 'Osvaldo', 'Ottavio', 'Ottone', 'Pandulf', 'Panfilo', 'Paolo', 'Paride', 'Pasqual', 'Pasquale', 'Patrizio', 'Pellegrino', 'Peppino', 'Pier', 'Pierangelo', 'Piergiorgio', 'Piergiuseppe', 'Pierluigi', 'Piermaria', 'Piero', 'Pierpaolo', 'Piersanti', 'Pietro', 'Pompeo', 'Pomponio', 'Puccio', 'Raffaele', 'Raffaellino', 'Raffaello', 'Raimondo', 'Ranieri', 'Rembrandt', 'Renzo', 'Riccardo', 'Ricciotti', 'Rino', 'Roberto', 'Rocco', 'Rodolfo', 'Rolando', 'Roman', 'Romeo', 'Romolo', 'Ronaldo', 'Rosario', 'Ruggero', 'Ruggiero', 'Sabatino', 'Salvatore', 'Salvi', 'Sandro', 'Sante', 'Santino', 'Saverio', 'Sebastiano', 'Sergius', 'Severino', 'Silvestro', 'Silvio', 'Simone', 'Stefano', 'Tazio', 'Telemaco', 'Temistocle', 'Tiziano', 'Tommaso', 'Toni', 'Tonino', 'Tonio', 'Torquato', 'Tullio', 'Ubaldo', 'Uberto', 'Ugo', 'Ugolino', 'Umberto', 'Valerio', 'Venancio', 'Vincentio', 'Vincenzo', 'Virgilio', 'Vito', 'Vittorio', 'Vladimiro', 'Wladimiro', 'Zanobi']
last_names = ['Rossi', 'Russo', 'Ferrari', 'Esposito', 'Bianchi', 'Romano', 'Regio', 'Ricci', 'Marino', 'Lupo', 'Lastra', 'Bruno', 'Gallo', 'Conti', 'De Luca', 'Mancini', 'Costa', 'Giordano', 'Rizzo', 'Lombardi', 'Moretti']
middle = ' was a Venetian '
occupations = ['Acrobat', 'Alchemist', 'Apothecarist', 'Architect', 'Astrologer', 'Armorer', 'Artist', 'Baker', 'Barrister', 'Bookbinder', 'Bowyer', 'Basket Weaver', 'Blacksmith', 'Brewer', 'Brick Layer', 'Butcher', 'Calligrapher', 'Candlemaker', 'Carpenter', 'Cartographer', 'Charcoal Burner', 'Clerk', 'Clothier', 'Cook', 'Coppersmith', 'Cooper', 'Diplomat', 'Dyer', 'Engineer', 'Engraver', 'Falconer', 'Farmer', 'Fisherman', 'Fishmonger', 'Forester', 'Fortune-Teller', 'Fruitier', 'Fuller', 'Furrier', 'Glassblower', 'Goldsmith', 'Grocer', 'Gardener', 'Grain Merchant', 'Grave Digger', 'Haberdasher', 'Herald', 'Herbalist', 'Hunter', 'Inkeeper', 'Interpreter', 'Jester', 'Jeweler', 'Lacemaker', 'Leatherworker', 'Locksmith', 'Mason', 'Mercer', 'Miller', 'Minstrel', 'Messenger', 'Miner', 'Moneylender', 'Navigator', 'Needleworker', 'Painter', 'Pardoner', 'Peddler', 'Priest', 'Physician', 'Playwright', 'Politician', 'Potter', 'Rat Catcher', 'Sailor', 'Scribe', 'Servant', 'Shipwright', 'Shoemaker', 'Silversmith', 'Solicitor', 'Soapmaker', 'Stoncarver', 'Storyteller', 'Spy', 'Tanner', 'Towne Crier', 'Vintner', 'Washer Woman', 'Waterman', 'Weaver', 'Wet Nurse', 'Wheelwright Wood Carver', 'Woodworker']
end = '. [SEP] '

def generate_random_fake_bio():
    import random

    prompt = random.choice(first_names) + ' ' + random.choice(last_names) + middle + random.choice(lower(occupations)) + end
    return generate_fake_bio(prompt)

In [8]:
print(generate_random_fake_bio())

Gavino Lastra was a Venetian Storyteller. [SEP]    He was born in Venice, the son of Tommaso and Amelia Gordino. Tommaso began his story-telling at an early age. One of his first jobs was as a scribe for Cardinal Francesco Barberini. Tommaso then worked for the court of Cardinal Francesco Baldissera Bartolomeo Cornaro and finally for Cardinal Francesco Sixtus IV.    Tommaso developed a style which was very much in keeping with the literary standards of his time. His diction was often accurate, but his sentences often convoluted or garbled. His syntax was also very much in keeping with the times, having become much more complex than it was before. He used many compositional devices, and his plots often felt as though they were entirely planned out.    Tommaso first gained prominence as a writer in the mid-18th century.  Between the years 1775 and 1805, he wrote several manuscripts, mostly on religious and mythological topics, for the Venetian Academy. These works were edited by Francesc

In [ ]:
for _ in range(1200):
    string = generate_random_fake_bio()
    with open('fakes.csv', 'a', encoding='utf-8') as file:
        file.write(f"{string}\n")

Rosario Rizzo was a Venetian Servant. [SEP]      He was born to a Venetian family in Padua, the son of Giambattista Rizzo and his wife Cecilia Contarini. His father was Giovanni Rizzo, a lackey of the Medici family, and his mother was Cecilia Morosini. He was educated at Padua and received at Padua the higher mathematics and humanist education. He was admitted to the University of Padua in 1717.He is believed to have accompanied his father to Florence in 1717, but in that year moved to Venice, to work as a tutor for his brother Bernardo.    In 1718, he took a position as a teacher at the University of Florence. On 2 July 1719, he accepted an appointment as a lecturer at the University of Padua, where he taught mathematics to Francesco Foscari, Giovanni Battista Piazzetta, and Antonio Sdzinski. In 1722, he was appointed as a professor of mathematics to the Università degli Università degli Studi e Latteschi in Venice, and he was a lecturer at the University of Padua the following year. 

Cirillo Lombardi was a Venetian Barrister. [SEP]    He was born in Venice, the son of Tommaso Lombardi and Theresa Novelli. Tommaso was a man of humble origin, born into a family of average wealth. Theresa Lombardi's father was a handicrafts merchant who did not want his son to pursue a career in Venice. Tommaso was not educated, and had to rely on his wife's knowledge. He began his legal studies at the University of Padua, where he obtained a law degree. In 1714, he entered the Venetian Senate, where he was the most senior member and the most influential. In his previous positions, he had to work for the Republic of Venice, because the Republic of Venice did not recognize the position of Senatearian. He was the Senatearian for the Serenissima, and later for the Republic of Genoa. After the revolution of 1618, he acted as treasurer to the Serenissima. He had to repay that relationship with Genoa, however, because of the actions of his predecessor, the Serenissima became allied with the

Ciro Bianchi was a Venetian Cartographer. [SEP]   
Giovanni Lastra was a Venetian Candlemaker. [SEP]     Giovanni Lastra was born in Venice on January 1, 1513. He began his apprenticeship with Andrea Mantegna, who would be his master for the next thirty years. The origin of his style is unclear; his work appears to stem from a late 15th-century design, but was influenced by various design traditions of the time. For his own part, Mantegna produced numerous illuminated manuscripts, with titles such as Santo Domini Lamentatura,  Lamentation on the  Destruction  of the  Temple of Solomon   and  La fanti nobili  (The Martyrdom of Solomon, liturgical text, liturgical text, and   title, with  illustrations).   
Delfino Conti was a Venetian Clothier. [SEP]    He was the son of Giacomo Dolfin and Elisabetta Conti. He was composer to the pianoforte by his uncle Giovanni Dolfin, and also wrote madrigals.    In Italy, he wrote two sets of twelve madrigals for four strings, published in 1824 and 1

Silvestro Gallo was a Venetian Storyteller. [SEP]    He was a librettist and masquerades as a clown in order to trick the people of Venice into believing he is actually a librettist. He also changes his appearance depending on the mood of the moment. His libretti are humorous in nature, and often have a dramatic structure, so that the audience is left with the feeling that the author is actually amused. He is considered to be one of the three great storytellers of Venice, alongside Francesco Mosconi and Giovanni Francesco Bembo.    He died at Venice in  1703.   
Adamo De Luca was a Venetian Servant. [SEP]    He was born in Venice, the son of Giovanna Morlaiter and her husband, the Captain general of the Sea Francesco Morlaiter. He served as a soldier in the wars of the Bourbons and the Papacy, and participated in the negotiations that led to the Treaty of Piacenza. In 1694, he was appointed governor of Venice, and held that office until 1702, when he was elected as doge. He was a puppe

Piergiorgio Regio was a Venetian Weaver. [SEP]    He was born in Venice and was educated at the School of Orazio. He spent some time in Germany and France. He was in England by 1774. He was in England by 1796. He was in Germany by 1797. He was once again in England by 1812. He was in England by 1813. In 1814, he was in Scotland. In 1815, he was in France. He was in France by 1817. He was in Spain again in 1821. He was in England by 1827. He was in Spain again in 1828. He was in Italy by 1829. He was in Venice by 1832. He was in Italy by 1833. He was in England by 1837. He was in Scotland by 1838. He was in Scotland by 1839. He was in Italy by 1841. He was in England by 1844. He was in Italy by 1847. He was in Germany by 1849. He was in Germany by 1851. He was in Italy by 1854. In 1856, he was in England. He was in Germany by 1859. He returned to Spain in 1867.     He is also known as   Pardo de la Fornarina, Pardo dal Pardo, or La Fornarina. The Spanish translation of his name is unkno

Costantino Mancini was a Venetian Minstrel. [SEP]    He was born in Venice, belonged to the nobility of the city. He was a seeker after truth, in the service of the republic, and was initiated into the mysteries of the academy of Bologna. In 1597 he was professor of mathematics at the University of Padua, and in 1602, member of the Academy of Genoa. In 1612, he became professor in Padua, and was again appointed professor in 1615, for which time he was awarded a large salary and valuable employments. During the course of his time in Padua, he became acquainted with  the vineyard industry of Padua, and there collaborated in the improvement of the art of vine-garden.    In 1615, he became a chancellor (presidentium generale) of the University of Padua. In 1617, he became a member of the Council of Ten. In 1626, he became a member of the Council of Twelve. In 1630, he became a member of the Audit Council of Rome. In 1634, he became provveditore generale di Nivo. In 1636, he became a member

Cesare Mancini was a Venetian Jeweler. [SEP]    He was the son of Silvestro Mancini, a silversmith. Silvestro married Francesco Barbaro, the daughter of Giovanni Battista Barbaro, who was also a Venetian.  Silvestro had two sons, Giacomo and Francesco, who were all killed in the plague.    Mancini's business was failing, so he and his brother Giacomo sold their share of the shop in the beginning of 1531, when the Venetian Council of Ten (Consiglio della Governatore Governatore) confiscated their interests in order to support the faltering Venetian economy.  In 1535, Mancini became the first owner of a share of the new condominium building The Athenaeum, which was inaugurated in 1538. The building was designed by architect Vincenzo Scamozzi and built in the façade of the old Chiesa Palace in Venice with Venetian assistance.      In 1537, Mancini, along with his brother Giacomo, sold their share of the condominium building  The Athenaeum to Francesco Barbaro, the father of Andrea Barbaro

Nico Bruno was a Venetian Gardener. [SEP]    He was born in Venice. He was a Venetian resident and the gardener at the Villa Imperiale. He was married to Caterina Gozzadini. He spent some time in Paris. He died in Venice.   
.
Jacopo Romano was a Venetian Pardoner. [SEP]    He was born on the island of Rab, in the far north of Dalmatia, from a Venetian family that produced many more Pardons than Palazzi. He was the son of the noted Venetian painter, Stefano Romano. His brother, Marco Romano also painted. He attended the Academy of Fine Arts of Venice, where he won a prize for painting a Virgin and Saints.    He was active in Venetian painting in the early years of the 18th century, and produced some important works there.    He was a prolific artist, and produced many important paintings in the Venetian school. Among his pupils were Francesco Foscari, Agostino Longhi, Ludovico Lipparini, Paolo Veronese, Pietro Netto, Giacomo Brunetti, and Antonio Molinari.       
Enzio Lastra was a Ven

Giacomo Russo was a Venetian Artist. [SEP]    He was probably born in Venice in 1490 or 1491, and trained with a Venetian painter named Francesco Moschini. He worked alongside Giovanni Antonio Fasolo in Venice in the 1520s and was an accomplice of Giovanni Antonio Maria Pieruzzi in Milan. In Milan he worked alongside Giovanni Battista Tiepolo. He painted a Madonna della Morte for the church of San Francesco della Vigna in Venice. He painted a Madonna del Roseto for the church of San Moisè in Venice. He painted an Adoration of the Magi for the Villa Pisani at Stra. He painted a Madonna Divini at the Villa Pisani at Stra. He painted an Adoration of the Princes at the Villa Pisani at Galipo La Furia. He painted a Martyrdom of Saint Romuald for the church of the Palazzo Mutazzo in Venice. He painted a Martyrdom of Saint Romuald for the church of the Padri Croce in Venice. He painted a Martyrdom of the Saints of Aversa and Angelo in Ravenna (1498) for the church of the Padri Croce in Venice

Giuseppe Mancini was a Venetian Shipwright. [SEP]     Marco Polo was born in Venice in 1516 as the son of a rich family. He was the first of six children, and his family was extremely proud of him. He was assigned a noble surname, and nicknamed "Pompeo", because of his poor breeding. His parents were both engravings and engravers. Polo was very interested in architecture, and as a boy he built and decorated small churches and chapels. He was also a bookseller and printed books. His family was always helpful to him, and he spent many years in various stages of study at the Academy of Urbino, and at the Academy of Lazzaro Bastiani. In 1566 he married Elisabetta Contarini, with whom he had three children: Andrea (1571–1594), Aldobrandino (1582–1634), and Giovanni.      In 1580, he was appointed Captain General of the Sea, in the ranks of the distinguished, and very wealthy, members of his family. It was a great honour, and well-known of his family that he was elected as a Colonel of the R

Salvatore Gallo was a Venetian Fuller. [SEP]    Born in Venice in the 1490s, Alvise Gallo spent much of his life in the city of Venice. At age 63, he decided to move to France, and in 1623, he served as a court physician to Louis I. From 1626 to 1629, he was a court physician to Charles I, with whom he was very close friends. As a physician, Gallo was very influential.    In the late 15th century, French physician and scholar Blaise Zoccola published several documents on the medical practice of Galba. These included the Magistracy of Galba, or, A Course of Medical Science; or, A Course of Practical Karate; or, Two Discourses on the Art of Fighting.    In 1633, Gallo published the Magistracy of Galba, which contained valuable historical, scientific, and moral information about the Roman Curia, the Byzantine Empire, and the medieval forms of medicine. Later, he published the  Jesu Christi generale, or, Jesu Christ and the Medico-Historical,   which contains historical material and a bibl

Santino Costa was a Venetian Bowyer. [SEP]    He was the son of Giovanni Battista Costa, a Venetian nobleman and the nephew of Andrea Costa, the first lord of Santorini. In the late 1490s, he was appointed bailo of Negroponte. As bailo, he negotiated a treaty to end the Second Crusade. In 1499, the Venetians secured approval from the League of Cambrai for the construction of the first ever canal shipping in the Mediterranean Sea. In 1508, he was appointed bailo of Rennes. He returned to Negroponte as bailo of Brixen in 1509. In 1509, he was appointed bailo of Tenos. He was also the bailo of Brixen from 1512 to 1516 and again from 1516 to 1521. In 1521, he was appointed bailo of Marseille. He was the bailo of Negroponte from 1523 to 1531. In 1526, he was appointed bailo of Lefkada. He became duke of Candia in 1527 and later duke of Candia, Della Rovere and now Duke of Candia.    In 1563, he was appointed bailo of Negroponte.In 1566, he was appointed bailo of Ventimiglia.In 1568, after t

Rodolfo Bruno was a Venetian Farmer. [SEP]    He was born in Venice in 1520. He was the son of Bartolomeo and Lucretia. His father died in 1531, leaving his mother Lucretia to assume the leadership of the household. In 1533, Bartolomeo and Lucretia married Ursus IV of Habsburg, Duke of Brunswick-Lüneburg.    In 1542, Bartolomeo participated in the War of Ferrara as a military commander in the Duke of Modena's army. In 1546, he was appointed governor of Fiesole.    In 1556, he was appointed governor of Ferrara again as a governor of warring tribes on the Ionian Sea. In 1564, he was appointed governor of Bologna.  In 1567, he was appointed governor of Vicenza.  In 1571, he was appointed governor of Padua.  In 1575, he was appointed governor of Vicenza.  In 1579 he was appointed governor of Bergamo.    Bartolomeo participated in the failed revolt of 1585–1586 under the leadership of Giovanni I Cornaro, Duke of Urbino, and Niccolò III Sorgi, governor of the Federici family.  Later that yea

In [ ]:
# fakes = []
# for i in range(1200):
#     fakes.append(generate_random_fake_bio())
#     if i % 50 == 0:
#         print('.')

# fakes_df = pd.DataFrame(fakes, columns=["bio"])
# fakes_df.to_csv('fakes.csv', index=False)
# # filename = 'fakes.csv'
# # shutil.copyfile(file_name, "/content/drive/Shared drives/FDH/data/"+file_name)

.


In [ ]:
fakes[0]

In [ ]:
fakes

In [4]:
# file_name='venetian_bios_2.csv'
# file_name='bios_old_dates.csv'
# shutil.copyfile("/content/drive/Shared drives/FDH/data/" + file_name, file_name)


df = pd.read_csv('bios_old_dates.csv')
df.columns = ['bio']
df['bio'].iloc[0]

'Rabbi Samuel ben Abraham Aboab (Hebrew:     ; c.\u20091610 – August 22, 1694) also known by his acronym RaSHA (", Rabbi Shmuel ben Avraham) was a 17th century Western Sephardic rabbi and scholar, who is considered to be one of the greatest rabbinic sages of Italy. [SEP] He served as the av bet din of Venice, where he rose to great prominence due to his vast knowledge of rabbinic literature. He is known for being an adamant opponent of the Sabbatean movement, and an early supporter of the old Yishuv.    Born into the Aboab family in Hamburg, Germany in 1610. His father Abraham Aboab V, was a former Crypto-Jew and founding father of the Portuguese Jewish community in Hamburg. At age 13, Rabbi Samuel was sent by his father to study in Venice under Rabbi David Franco, whose daughter, Mazzal-Tov Franco he later married at age eighteen. He was soon appointed Chief Rabbi of Verona, where he garnered such a reputation for learning that disciples sought him out, such as Samson Morpurgo and Jac

In [5]:
# add label column to train data (which is legit, so always 1)
df['label'] = 1

# remove non alphanumeric characters from the 'text' column
import re
def clean_txt(text):
    text = re.sub("'", "",text)
    text=re.sub("(\\W)+"," ",text)    
    return text

df['text'] = df['bio'].apply(clean_txt)

In [9]:
fakes_df = pd.read_csv('fakes.csv', sep='|')
fakes_df.columns = ['bio']
fakes_df['bio'].iloc[0]

'Constanzo Rossi was a Venetian Engineer. [SEP]    He was born in Venice in 1583, to Giacomo Rossi and Laura Scrini. He joined the Navy in 1809, at the age of 21, and fought in the Mediterranean Expedition in the War of the Austrian Succession. In 1811 he was appointed Professor of Propulsion and Propulsion Engineering at the Academy of Science of Berlin. In 1813 he was appointed Director of Propulsion and Propulsion Research at the Foundry of Erfurt am Sonntag in 1814, and was a member of the scientific advisory council of the Elector Institute. In 1816, he was appointed Professor of Propulsion and Propulsion Engineering at the University of Vienna. In 1818, he was appointed a Fellow of the Royal Society in London. In 1822, he was appointed Vice-Chancellor of the Royal Society in Edinburgh. In 1824, he was elected a member of the Royal Society of London.    In 1810, he and his brother Marino completed the first prototype of the "superconductor", a barrel composed of several condensers

In [10]:
# fakes = []
# for i in ["100", "200", "400", "600", "800"]:
#     fakes.append(pd.read_csv(f"bert_gan_fake{i}.csv"))
    
# fake = pd.concat(fakes)
# fakes_df = pd.DataFrame({'text' : fakes})
fakes_df['text'] = fakes_df['bio'].apply(clean_txt)
fakes_df['label'] = 0
df = pd.concat([df, fakes_df])[['text', 'label']].dropna()

In [11]:
# split into test and val
from sklearn.model_selection import train_test_split
train, val = train_test_split(df, test_size=0.2, random_state=35)

train.reset_index(drop=True, inplace=True)
val.reset_index(drop=True, inplace=True)

print("Training Set Shape :", train.shape)
print("Validation Set Shape :", val.shape)

Training Set Shape : (1008, 2)
Validation Set Shape : (252, 2)


In [55]:
OUTPUT_DIR = 'tmp'

#@markdown Whether or not to clear/delete the directory and create a new one
DO_DELETE = False #@param {type:"boolean"}

if DO_DELETE:
    try:
        tf.gfile.DeleteRecursively(OUTPUT_DIR)
    except:
        pass

tf.gfile.MakeDirs(OUTPUT_DIR)
print('***** Model output directory: {} *****'.format(OUTPUT_DIR))

***** Model output directory: tmp *****


In [13]:
DATA_COLUMN = 'text'
LABEL_COLUMN = 'label'

# label_list is the list of labels, i.e. True, False or 0, 1
label_list = [0, 1]

In [14]:
def get_split(text1):
    l_total = []
    l_parcial = text1.split()[:200]
    l_total.append(" ".join(l_parcial))
    
    w = 1
    n = len(text1.split())
    while (200 + (w-1)*150) < n:
        l_parcial = text1.split()[ w*150 : (w*150 + 200) ]
        l_total.append(" ".join(l_parcial))
        w += 1
    return l_total

In [15]:
train['text_split'] = train[DATA_COLUMN].apply(get_split)
val['text_split'] = val[DATA_COLUMN].apply(get_split)

/dhlabdata4/bborges/.conda/envs/bertp2/lib/python3.6/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.
/dhlabdata4/bborges/.conda/envs/bertp2/lib/python3.6/site-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  



Go from array of chunks into dataframe of chunks.

In [16]:
train_l = []
label_l = []
index_l =[]
for idx,row in train.iterrows():
    for l in row['text_split']:
        train_l.append(l)
        label_l.append(row['label'])
        index_l.append(idx)

train_df = pd.DataFrame({DATA_COLUMN:train_l, LABEL_COLUMN:label_l})
train_df.head()

text  label
0  Luisa Bergalli 1703 in Venice 1779 in Venice w...      1
1  Battista Franco Veneziano also known by his co...      1
2  Antonio Diedo 15 November 1772 1 January 1847 ...      1
3  the Academy of Fine Arts of Venice In 1838 he ...      1
4  St SEP Mark Enthroned is an early painting by ...      1

In [17]:
val_l = []
val_label_l = []
val_index_l = []
for idx,row in val.iterrows():
    for l in row['text_split']:
        val_l.append(l)
        val_label_l.append(row['label'])
        val_index_l.append(idx)

val_df = pd.DataFrame({DATA_COLUMN:val_l, LABEL_COLUMN:val_label_l})
val_df.head()

text  label
0  Nicolò Contarini 26 September 1553 2 April 163...      1
1  Upon the death of Cornaro a heated debate led ...      1
2  Leonardo Donà or Donato Venice February 12 153...      1
3  future Pope Paul V Donatos staunchly anti papa...      1
4  1605 two weeks before Grimanis death Pope Paul...      1

Text pre-processing truly finished; start moving data onto BERT.

In [18]:
train_InputExamples = train_df.apply(lambda x: bert.run_classifier.InputExample(guid=None,
                                                                   text_a = x[DATA_COLUMN], 
                                                                   text_b = None, 
                                                                   label = x[LABEL_COLUMN]), axis = 1)

val_InputExamples = val_df.apply(lambda x: bert.run_classifier.InputExample(guid=None, 
                                                                   text_a = x[DATA_COLUMN], 
                                                                   text_b = None, 
                                                                   label = x[LABEL_COLUMN]), axis = 1)

In [19]:
# This is a path to an uncased (all lowercase) version of BERT
BERT_MODEL_HUB = "https://tfhub.dev/google/bert_uncased_L-12_H-768_A-12/1"

def create_tokenizer_from_hub_module():
    """Get the vocab file and casing info from the Hub module."""
    with tf.Graph().as_default():
        bert_module = hub.Module(BERT_MODEL_HUB)
        tokenization_info = bert_module(signature="tokenization_info", as_dict=True)
        with tf.Session() as sess:
            vocab_file, do_lower_case = sess.run([tokenization_info["vocab_file"],
                                            tokenization_info["do_lower_case"]])
      
    return bert.tokenization.FullTokenizer(
      vocab_file=vocab_file, do_lower_case=do_lower_case)

tokenizer = create_tokenizer_from_hub_module()

In [20]:
MAX_SEQ_LENGTH = 200

# Convert our train and validation features to InputFeatures that BERT understands.
train_features = bert.run_classifier.convert_examples_to_features(
    train_InputExamples, label_list, MAX_SEQ_LENGTH, tokenizer)

val_features = bert.run_classifier.convert_examples_to_features(
    val_InputExamples, label_list, MAX_SEQ_LENGTH, tokenizer)

In [56]:
def create_model(is_predicting, input_ids, input_mask, segment_ids, labels,
                                 num_labels):
    """Creates a classification model."""

    bert_module = hub.Module(
            BERT_MODEL_HUB,
            trainable=True)
    bert_inputs = dict(
            input_ids=input_ids,
            input_mask=input_mask,
            segment_ids=segment_ids)
    bert_outputs = bert_module(
            inputs=bert_inputs,
            signature="tokens",
            as_dict=True)

    # Use "pooled_output" for classification tasks on an entire sentence.
    # Use "sequence_outputs" for token-level output.
    output_layer = bert_outputs["pooled_output"]

    hidden_size = output_layer.shape[-1].value

    # Create our own layer to tune for politeness data.
    output_weights = tf.get_variable(
            "output_weights", [num_labels, hidden_size],
            initializer=tf.truncated_normal_initializer(stddev=0.02))

    output_bias = tf.get_variable(
            "output_bias", [num_labels], initializer=tf.zeros_initializer())

    with tf.variable_scope("loss"):

        # Dropout helps prevent overfitting
        output_layer = tf.nn.dropout(output_layer, keep_prob=0.5)

        logits = tf.matmul(output_layer, output_weights, transpose_b=True)
        logits = tf.nn.bias_add(logits, output_bias)
        log_probs = tf.nn.log_softmax(logits, axis=-1)

        # Convert labels into one-hot encoding
        one_hot_labels = tf.one_hot(labels, depth=num_labels, dtype=tf.float32)

        predicted_labels = tf.squeeze(tf.argmax(log_probs, axis=-1, output_type=tf.int32))
        # If we're predicting, we want predicted labels and the probabiltiies.
        if is_predicting:
            return (predicted_labels, log_probs)

        # If we're train/eval, compute loss between predicted and actual label
        per_example_loss = -tf.reduce_sum(one_hot_labels * log_probs, axis=-1)
        loss = tf.reduce_mean(per_example_loss)
        return (loss, predicted_labels, log_probs)


# model_fn_builder actually creates our model function
# using the passed parameters for num_labels, learning_rate, etc.
def model_fn_builder(num_labels, learning_rate, num_train_steps,
                                         num_warmup_steps):
    """Returns `model_fn` closure for TPUEstimator."""
    def model_fn(features, labels, mode, params):    # pylint: disable=unused-argument
        """The `model_fn` for TPUEstimator."""

        input_ids = features["input_ids"]
        input_mask = features["input_mask"]
        segment_ids = features["segment_ids"]
        label_ids = features["label_ids"]

        is_predicting = (mode == tf.estimator.ModeKeys.PREDICT)
        
        # TRAIN and EVAL
        if not is_predicting:

            (loss, predicted_labels, log_probs) = create_model(
                is_predicting, input_ids, input_mask, segment_ids, label_ids, num_labels)

            train_op = bert.optimization.create_optimizer(
                    loss, learning_rate, num_train_steps, num_warmup_steps, use_tpu=False)

            # Calculate evaluation metrics. 
            def metric_fn(label_ids, predicted_labels):
                accuracy = tf.metrics.accuracy(label_ids, predicted_labels)
                f1_score = tf.contrib.metrics.f1_score(
                        label_ids,
                        predicted_labels)
                auc = tf.metrics.auc(
                        label_ids,
                        predicted_labels)
                recall = tf.metrics.recall(
                        label_ids,
                        predicted_labels)
                precision = tf.metrics.precision(
                        label_ids,
                        predicted_labels) 
                true_pos = tf.metrics.true_positives(
                        label_ids,
                        predicted_labels)
                true_neg = tf.metrics.true_negatives(
                        label_ids,
                        predicted_labels)     
                false_pos = tf.metrics.false_positives(
                        label_ids,
                        predicted_labels)    
                false_neg = tf.metrics.false_negatives(
                        label_ids,
                        predicted_labels)
                return {
                        "eval_accuracy": accuracy,
                        "f1_score": f1_score,
                        "auc": auc,
                        "precision": precision,
                        "recall": recall,
                        "true_positives": true_pos,
                        "true_negatives": true_neg,
                        "false_positives": false_pos,
                        "false_negatives": false_neg
                }

            eval_metrics = metric_fn(label_ids, predicted_labels)

            if mode == tf.estimator.ModeKeys.TRAIN:
                return tf.estimator.EstimatorSpec(mode=mode,
                    loss=loss,
                    train_op=train_op)
            else:
                    return tf.estimator.EstimatorSpec(mode=mode,
                        loss=loss,
                        eval_metric_ops=eval_metrics)
        else:
            (predicted_labels, log_probs) = create_model(
                is_predicting, input_ids, input_mask, segment_ids, label_ids, num_labels)

            predictions = {
                    'probabilities': log_probs,
                    'labels': predicted_labels
            }
            return tf.estimator.EstimatorSpec(mode, predictions=predictions)

    # Return the actual model function in the closure
    return model_fn

In [57]:
BATCH_SIZE = 16
LEARNING_RATE = 2e-5
NUM_TRAIN_EPOCHS = 2.0
# Warmup is a period of time where the learning rate is small and gradually increases--usually helps training.
WARMUP_PROPORTION = 0.1
# Model configs
SAVE_CHECKPOINTS_STEPS = 300
SAVE_SUMMARY_STEPS = 100

# Compute train and warmup steps from batch size
num_train_steps = int(len(train_features) / BATCH_SIZE * NUM_TRAIN_EPOCHS)
num_warmup_steps = int(num_train_steps * WARMUP_PROPORTION)

# Specify output directory and number of checkpoint steps to save
run_config = tf.estimator.RunConfig(
    model_dir=OUTPUT_DIR,
    save_summary_steps=SAVE_SUMMARY_STEPS,
    save_checkpoints_steps=SAVE_CHECKPOINTS_STEPS)

# Specify output directory and number of checkpoint steps to save
run_config = tf.estimator.RunConfig(
    model_dir=OUTPUT_DIR,
    save_summary_steps=SAVE_SUMMARY_STEPS,
    save_checkpoints_steps=SAVE_CHECKPOINTS_STEPS)

In [58]:
#Initializing the model and the estimator
model_fn = model_fn_builder(
  num_labels=len(label_list),
  learning_rate=LEARNING_RATE,
  num_train_steps=num_train_steps,
  num_warmup_steps=num_warmup_steps)

estimator = tf.estimator.Estimator(
  model_fn=model_fn,
  config=run_config,
  params={"batch_size": BATCH_SIZE})

In [59]:
# Create an input function for training. drop_remainder = True for using TPUs.
train_input_fn = bert.run_classifier.input_fn_builder(
    features=train_features,
    seq_length=MAX_SEQ_LENGTH,
    is_training=True,
    drop_remainder=False)

test_input_fn = run_classifier.input_fn_builder(
    features=val_features,
    seq_length=MAX_SEQ_LENGTH,
    is_training=False,
    drop_remainder=False)

In [60]:
# Training the model
print(f'Beginning Training!')
current_time = datetime.now()
estimator.train(input_fn=train_input_fn, max_steps=num_train_steps)
print("Training took time ", datetime.now() - current_time)

Beginning Training!


/dhlabdata4/bborges/.conda/envs/bertp2/lib/python3.6/site-packages/tensorflow/python/ops/gradients_util.py:93: UserWarning: Converting sparse IndexedSlices to a dense Tensor of unknown shape. This may consume a large amount of memory.
  "Converting sparse IndexedSlices to a dense Tensor of unknown shape. "


Training took time  0:14:43.161784


In [61]:
# Evaluating the model with Validation set
estimator.evaluate(input_fn=test_input_fn, steps=None)

/dhlabdata4/bborges/.conda/envs/bertp2/lib/python3.6/site-packages/tensorflow/python/ops/gradients_util.py:93: UserWarning: Converting sparse IndexedSlices to a dense Tensor of unknown shape. This may consume a large amount of memory.
  "Converting sparse IndexedSlices to a dense Tensor of unknown shape. "


{'auc': 0.866433,
 'eval_accuracy': 0.9705015,
 'f1_score': 0.98439926,
 'false_negatives': 11.0,
 'false_positives': 9.0,
 'loss': 0.09608859,
 'precision': 0.9859375,
 'recall': 0.98286605,
 'true_negatives': 27.0,
 'true_positives': 631.0,
 'global_step': 299}


Model trained, select best chunks.

In [62]:
def getPrediction(in_sentences):
    labels = ["Fake", "Real"]
    input_examples = [run_classifier.InputExample(guid="", text_a = x[0], text_b = None , label = 0) 
                      for x in in_sentences] # here, "" is just a dummy label
    input_features = run_classifier.convert_examples_to_features(
        input_examples, label_list, MAX_SEQ_LENGTH, tokenizer)
    predict_input_fn = run_classifier.input_fn_builder(
        features=input_features, seq_length=MAX_SEQ_LENGTH, is_training=False, drop_remainder=False)
    print(predict_input_fn)
    predictions = estimator.predict(predict_input_fn)
    return pd.DataFrame([(sentence, np.exp(prediction['probabilities'][1]), 
                          labels[prediction['labels']]) 
                         for sentence, prediction in zip(in_sentences, predictions)], 
                        columns=['bio', 'prob_real','label'])

In [63]:
real_sent = val.loc[val['label']==1,['text']].values.tolist()

predictions_real = getPrediction(real_sent)
predictions_real.sort_values('prob_real', ascending=False)

<function input_fn_builder.<locals>.input_fn at 0x7fa180aaa378>


bio  prob_real label
39   [Pierre Spitznagel O Carm SEP 1411 1465 was a ...   0.999395  Real
125  [The Two Foscari An Historical Tragedy 1821 is...   0.999278  Real
47   [Theodora Doukaina Greek before 1059 after 107...   0.999166  Real
1    [Leonardo Donà or Donato Venice February 12 15...   0.999157  Real
2    [Girolamo Michelangelo Grigoletti 29 August 18...   0.999079  Real
..                                                 ...        ...   ...
88   [Nicolao Colletti also written Coletti 18th ce...   0.984484  Real
108  [Ugo Pendini August 14 1853 1895 was an Italia...   0.983787  Real
31   [Albano Contarini was a Venetian nobleman and ...   0.980954  Real
217  [Lorenzo de Monacis was a distinguished diplom...   0.049343  Fake
169  [Contarina Barbarigo died 25 December 1804 was...   0.025767  Fake

[237 rows x 3 columns]

In [49]:
# proposed_replies = pd.read_csv('TODO_proposed_replies.csv').dropna()
# replies_to_test = proposed_replies[['text']].values.tolist()
proposed_bio = generate_fake_bio("Antonio Di Wolfio was a venetian hunter. [SEP] ")

In [50]:
predictions_proposed = getPrediction(get_split(proposed_bio) + ['dummy'])
predictions_proposed.sort_values('prob_real', ascending=False)

<function input_fn_builder.<locals>.input_fn at 0x7fa180d9ce18>


bio  prob_real label
4  with the title of "superintended" bishop of Ve...   0.210850  Fake
2  1431 to 1432, he was lord of Negroponte. On 7 ...   0.195992  Fake
1  to the Italian throne, resisted by the Republi...   0.185716  Fake
5                                              dummy   0.184486  Fake
0  Antonio Di Wolfio was a venetian hunter. [SEP]...   0.146085  Fake
3  Saints Nazario and Celso in Vicenza. On 4 Janu...   0.135605  Fake

In [51]:
predictions_proposed = getPrediction(get_split(clean_txt(proposed_bio)) + ['dummy'])
predictions_proposed.sort_values('prob_real', ascending=False)

<function input_fn_builder.<locals>.input_fn at 0x7fa3e9927f28>


bio  prob_real label
1  with the succession to the Italian throne resi...   0.205483  Fake
4  cardinal by Pope Paul IV with the title of sup...   0.200745  Fake
5                                              dummy   0.181982  Fake
3  abbot of the Abbey of Saints Nazario and Celso...   0.162928  Fake
0  Antonio Di Wolfio was a venetian hunter SEP He...   0.150778  Fake
2  of Chios and from 1431 to 1432 he was lord of ...   0.068841  Fake

In [52]:
proposed_bio

'Antonio Di Wolfio was a venetian hunter. [SEP]    He was a member of the famous Wolfi family of Venetians. His brother Benedetto returned home. He was a member of a noble family from Bergamo, now in decline. He had two sons: Benedetto III d\'Este (1490–1511) and Benedetto IV (1511–1553). His daughters Elena (d. 1572) and Marietta (d. 1694) had two daughters who were all  prominent in Venetian politics and culture. His son Bernardo IV (1554–1629) was a very worthy opponent of the Bernardo family, as he was a descendant of the Lombard traders and had close ties with the Bolognese.    He was one of the richest Venetian patricians. He owned seven chancery in Venice, one of the most important estates of the Venetian patriciate.    On 14 January 1411, Antonio di Matteo married his cousin Caterina Barbarigo, daughter of the noble dukes of Urbino and Titian (antagonistic unions with the succession to the Italian throne, resisted by the Republic of Venice).    On 13 January 1412, he was electe

In [53]:
predictions_proposed = getPrediction(get_split(clean_txt("""Federico Faggin (born 1 December 1941) is an Italian-American physicist, engineer, inventor and entrepreneur. He is best known for designing the first commercial microprocessor, the Intel 4004. He led the 4004 (MCS-4) project and the design group during the first five years of Intel's microprocessor effort. Faggin also created, while working at Fairchild Semiconductor in 1968, the self-aligned MOS (metal–oxide–semiconductor) silicon-gate technology (SGT), which made possible MOS semiconductor memory chips, CCD image sensors, and the microprocessor. After the 4004, he led development of the Intel 8008 and 8080, using his SGT methodology for random logic chip design, which was essential to the creation of early Intel microprocessors. He was co-founder (with Ralph Ungermann) and CEO of Zilog, the first company solely dedicated to microprocessors, and led the development of the Zilog Z80 and Z8 processors. He was later the co-founder and CEO of Cygnet Technologies, and then Synaptics.

In 2010, he received the 2009 National Medal of Technology and Innovation, the highest honor the United States confers for achievements related to technological progress. In 2011, Faggin founded the Federico and Elvia Faggin Foundation to support the scientific study of consciousness at US universities and research institutes. In 2015, the Faggin Foundation helped to establish a $1 million endowment for the Faggin Family Presidential Chair in the Physics of Information at UC Santa Cruz to promote the study of "fundamental questions at the interface of physics and related fields including mathematics, complex systems, biophysics, and cognitive science, with the unifying theme of information in physics." Federico Faggin has been a Silicon Valley resident since 1968 and is a naturalized US citizen.""")) + ['dummy'])
predictions_proposed.sort_values('prob_real', ascending=False)

<function input_fn_builder.<locals>.input_fn at 0x7fa103d4c268>


bio  prob_real label
2                                              dummy   0.153999  Fake
0  Federico Faggin born 1 December 1941 is an Ita...   0.131954  Fake
1  of Cygnet Technologies and then Synaptics In 2...   0.090662  Fake

In [54]:
predictions_proposed = getPrediction(get_split(clean_txt("""Panfilo Castaldi (c. 1398 – c. 1490; also called Pamfilo or Pamphilo) was an Italian physician and "master of the art of printing," to whom local tradition attributes the invention of moveable type. He was born in Feltre but spent most of his life working in Milan.

The story as it has circulated through the centuries in Feltre is that Castaldi was given examples of early Chinese block printing by Marco Polo, with which he experimented, eventually producing modern type. The story was largely unknown outside of Lombardy until it was reported in the 19th century by Robert Curzon, Baron Zouche, a diplomat. As Curzon tells it, Castaldi began with glass stamps made at Murano and eventually developed wooden printing blocks which he used in a printing press in Venice in 1426. This would have been several years before Johann Gutenberg's first experiments with metal type in the early 1430s. Curzon stresses the connection to Marco Polo, arguing that Castaldi's (undated) early work closely resembles Chinese printing, and also stresses Gutenberg's acquaintance with Venetian printing, thus suggesting that European printing did not emerge distinctly from the Chinese analogues but in imitation (a frequently debated question in any case).

Since no more recent scholarship has suggested such an early date for Castaldi's printing, any relationship between Castaldi and Marco Polo, or any connection between Gutenberg and Northern Italy, the story has never received much credit outside of Feltre. In that city, however, it still attracts a fair amount of attention (the statue of Castaldi still stands) and was brought up once again on the occasion of Castaldi's sexcentennial in 1998; in a proposal for a municipal bill recognizing Castaldi, he was referred to as "il primo inventore dei caratteri mobili per la stampa" ("the first inventor of moveable type for printing").[6] What is clear is that by the 1470s Castaldi was a successful printer; there is a record of a print run of 300 copies of a Cicero epistle in 1471.[7]

Panfilo Castaldi is the imprint of a 19th and 20th-century printing house in Feltre.""")) + ['dummy'])
predictions_proposed.sort_values('prob_real', ascending=False)

<function input_fn_builder.<locals>.input_fn at 0x7fa103d4c488>


bio  prob_real label
1  stresses the connection to Marco Polo arguing ...   0.191589  Fake
2                                              dummy   0.165001  Fake
0  Panfilo Castaldi c 1398 c 1490 also called Pam...   0.092988  Fake

In [79]:
[print(s) for s in get_split(split_proposed_bio)]

Giovanni Domenico was a Venetian opera singer. [SEP] He sang at least 12 operas, including four repertory operas, four anthologies and six short, five-minute seriaticks. He is known to have developed the habit of singing with his right hand, which is distinctive to Venetian right hand gestures. He also adopted the habit of bringing his right hand to his mouth, which is unusual in Venetian right hand gestures. He sang in several Venetian dialects, including Italic, Lateran, and Venetian (land) songs. He wrote at least five published pieces in his native Venice, including several lengthy marches, shore marches, and patriotic songs in Italian. Giovanni Domenico was born in Venice on March 11, 1819. He is listed as 19 years old on a mercantile list of Venetian singers in 1878. He was a talented singer, who, thanks to social circumstances and education, came into prominence as a singer in Venice. He was the son of Bortolo Domenico, a singer from Mantua. In 1884, he married Francesca Morlait

[None, None, None, None]

In [33]:
proposed_bio

'Racism, as well as the related Problem of Difference and Inequality, was a prominent issue in the Italian Renaissance between the Renaissance and the turn of the 19th century.        The Renaissance were divided into two broad currents, the Neoclassical and the Neoclassical, and although some of their stylistic features resembled each other, they did not follow any particular conceptual or historical model. The main differences between the two, and their relationship to one another, are described in detail later in this article.      Identifying with the Neoclassical tradition, the Baroque ornament of Venetian stained glass is characterized by light golden and gold halo effects, and uses monograms in a complex hierarchical fashion. This style is similar to that of Giovanni Bellini, the influence of Ursus Rex, and references to the Byzantine art of the period. Among its works are  a  bas-relief of Santi Rota in the National Gallery of Art, Washington D.C.  a  divination by the Doge in 

In [ ]:
# best_predictions = (predictions_proposed.sort_values('prob_real', ascending=False)
#                     .groupby('comment').first().reset_index())

In [ ]:
# best_predictions.loc[best_predictions['prob_real']>0.9]#.to_csv('realistic_replies.csv')

Giovanni Domenico was a Venetian opera singer. [SEP] He sang in operas by Benedetto Ferrari, Giovanni Francesco Guardi, Francesco Hayez, the winner of the Golden Jubilee of Venice, and the composer of operas by Francesco Hayez, including 'Signor Paolo De Negri', 'La Fenice', and 'Derriere Ruprechtigo' as well as 'Te Classee Con Agnus Deo' and 'Derriere Ruprechtigo' as well as 'La Fenice' and 'Derriere Ruprechtigo' as well as 'De Confessione Dei Giovvani' and 'Confessione Dei Giovvani' as well as 'Confessione Mi Giovvani' and 'Confessione Mi vino Giovvani' as well as 'Te Classee Con Agnus Deo' and 'Confessione Mi Giovvani'.          Francesco Hayez (1527–1590) was an Italian librettist and opera composer. [SEP] He wrote libretti for operas by Francesco Hayez, including 'Faustina', 'Lambeth' and 'De Confessione'.             Francesco Hayez (fl. 1584–1590) was an Italian librettist and opera composer. [SEP] He wrote libretti for operas by Francesco Hayez, including 'Faustina', 'Lambeth' and 'Confessione'.        He was born in Venice, Italy in 1514. His father was Giovanni Francesco Hayez, a theatrical impresario who specialized in theatrical presentation of religious texts. His mother was Lucia Lombardo, sister of the composer Antonio Lombardo. He took classical education from his father. Hayez studied music with his uncle, and gained a taste for the finer points of the instrument. He was hired by the Camaldolese order in Camaldoli in 1545, and received their highest honor, the Camaldolese Cross of St. Michael holding the "Bishopric of Christ" (Cistercalium). In 1547 he visited Rome and became a novice abbot of the Benedictine monastery at Chiesa in Veneto. From 1548-50 he was appointed as head of the composers' school at the Johannist Ascension in Karlovci, and later served as chancellor of the Conservatory of the Triumph of Wisdom from 1551 to 1554. In 1555 he was appointed as director of the Gesuati (agricultural college) of Murano, and later as Dean of the School of Arts from 1565 to 1580.                 He composed several sacred music, including   his own motets (Mantaschinum verum 2, 1554, Venice),   his own  four works for four vocal orantioi (Mantaschinum verum 2, 1560, Venice), and  a motet for two (Mantaschinum verum 2, 1575, Venice).  He also wrote madrigals, canzonette, and canza (ritual for two flautini or four vocal or four flautini).             He died in Venice in 1590.         
